### ISSUE TITLE
with nn.DataParallel program runs slower

### ISSUE THREAD

**Issue by Apogentus** [pytorch/pytorch#5421]
Hello,
I ran classification tutorial from pytorch, and I noticed that using nn.DataParallel wrapper around the model forces forward&backward passes to run much slower. How can I take advantage of 2 GPUs?

I use Ubuntu 16.04 LTS, python 3.6, pytorch 0.3.1.post2, cuDNN 7005, 2 GPUs GeForce 1080.

Code:
```
%matplotlib inline

import torch
import torchvision
import torchvision.transforms as transforms

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=16,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=16,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')


from torch.autograd import Variable
import torch.nn as nn
import torch.nn.functional as F

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 20, 3)
        self.conv2 = nn.Conv2d(20, 40, 3)
        self.conv3 = nn.Conv2d(40, 80, 5)
        self.pool = nn.MaxPool2d(3, 3)
        self.fc1 = nn.Linear(5120, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = F.relu(self.conv3(x))
        x = self.pool(x)
        #print(x.shape)
        x = x.view(-1, 5120)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x

dataiter = iter(testloader)
inputs, labels = dataiter.next()
inputs, labels = Variable(inputs.cuda()), Variable(labels.cuda())

criterion = nn.CrossEntropyLoss()

net = Net().cuda()

%timeit loss=criterion(net(inputs),labels); loss.backward()

net = nn.DataParallel(Net().cuda())

%timeit loss=criterion(net(inputs),labels); loss.backward()
```
1st timeit gives 2 ms, second - 6ms, though second utilizes 2 GPUs.

**Comment by Kaixhin** [comment]
Your network looks relatively small, and I would assume that the overhead of moving data into two separate GPUs and then merging them back again is more than the gain you'd get from using two GPUs in parallel.

**Comment by apaszke** [comment]
Also, it’s not the right way to benchmark CUDA code. You should at least have a torch.cuda.synchronize() call at the end of the time it block, and do a small warmup first (a few passes to allocate memory etc)

**Comment by Apogentus** [comment]
Yes, increasing network depth and adding torch.cuda.synchronize()  gave a speed 1.5 improvement. Thank you.
